p2p/rating/peersRatingHandler.go

```diff
@@ -0,0 +1,251 @@
+package rating
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go-core/core"
+	"github.com/ElrondNetwork/elrond-go-core/core/check"
+	"github.com/ElrondNetwork/elrond-go-core/core/random"
+	"github.com/ElrondNetwork/elrond-go/dataRetriever"
+	"github.com/ElrondNetwork/elrond-go/p2p"
+	"github.com/ElrondNetwork/elrond-go/storage"
+)
+
+const (
+	topRatedTier   = "top rated tier"
+	badRatedTier   = "bad rated tier"
+	defaultRating  = 0
+	minRating      = -100
+	maxRating      = 100
+	increaseFactor = 2
+	decreaseFactor = -1
+	minNumOfPeers  = 1
+	int32Size      = 4
+)
+
+// ArgPeersRatingHandler is the DTO used to create a new peers rating handler
+type ArgPeersRatingHandler struct {
+	TopRatedCache storage.Cacher
+	BadRatedCache storage.Cacher
+	Randomizer    p2p.IntRandomizer
+}
+
+type peersRatingHandler struct {
+	topRatedCache storage.Cacher
+	badRatedCache storage.Cacher
+	randomizer    dataRetriever.IntRandomizer
+	mut           sync.Mutex
+}
+
+// NewPeersRatingHandler returns a new peers rating handler
+func NewPeersRatingHandler(args ArgPeersRatingHandler) (*peersRatingHandler, error) {
+	err := checkArgs(args)
+	if err != nil {
+		return nil, err
+	}
+
+	prh := &peersRatingHandler{
+		topRatedCache: args.TopRatedCache,
+		badRatedCache: args.BadRatedCache,
+		randomizer:    args.Randomizer,
+	}
+
+	return prh, nil
+}
+
+func checkArgs(args ArgPeersRatingHandler) error {
+	if check.IfNil(args.Randomizer) {
+		return p2p.ErrNilRandomizer
+	}
+	if check.IfNil(args.TopRatedCache) {
+		return fmt.Errorf("%w for TopRatedCache", p2p.ErrNilCacher)
+	}
+	if check.IfNil(args.BadRatedCache) {
+		return fmt.Errorf("%w for BadRatedCache", p2p.ErrNilCacher)
+	}
+
+	return nil
+}
+
+// AddPeer adds a new peer to the cache with rating 0
+// this is called when a new peer is detected
+func (prh *peersRatingHandler) AddPeer(pid core.PeerID) {
+	prh.mut.Lock()
+	defer prh.mut.Unlock()
+
+	_, found := prh.getOldRating(pid)
+	if found {
+		return
+	}
+
+	prh.topRatedCache.Put(pid.Bytes(), defaultRating, int32Size)
+}
+
+// IncreaseRating increases the rating of a peer with the increase factor
+func (prh *peersRatingHandler) IncreaseRating(pid core.PeerID) {
+	prh.mut.Lock()
+	defer prh.mut.Unlock()
+
+	oldRating, found := prh.getOldRating(pid)
+	if !found {
+		// new pid, add it with default rating
+		prh.topRatedCache.Put(pid.Bytes(), defaultRating, int32Size)
+		return
+	}
+
+	newRating := oldRating + increaseFactor
+	if newRating > maxRating {
+		return
```

Comment:
because of this condition we can have the following situation: oldRating is 99 and increaseFactor is 2, the current rating will be 99. Suggestion to change as following: 
```GO
    if newRating == maxRating {
        return
    }

    newRating := oldRating + increaseFactor
    if newRating > maxRating{
        newRating = maxRating     
    }
    
    prh.updateRating(pid, oldRating, newRating)
}
```
Valid also for DecreaseRating function

Reply:
fixed and moved to `updateRatingIfNeeded`